core/src/main/java/jenkins/security/apitoken/ApiTokenStats.java

```diff
@@ -0,0 +1,257 @@
+     * If the file is not present, a fresh new instance is created.
+     */
+    public static @Nonnull ApiTokenStats load(@Nonnull File parent) {
+        // even if we are not using statistics, we load the existing one in case the configuration
+        // is enabled afterwards to avoid erasing data
+        
+        XmlFile file = getConfigFile(parent);
+        ApiTokenStats apiTokenStats;
+        if (file.exists()) {
+            try {
+                apiTokenStats = (ApiTokenStats) file.unmarshal(ApiTokenStats.class);
+            } catch (IOException e) {
+                LOGGER.log(Level.WARNING, "Failed to load " + file, e);
+                apiTokenStats = new ApiTokenStats();
+            }
+        } else {
+            apiTokenStats = new ApiTokenStats();
+        }
+    
+        apiTokenStats.setParent(parent);
+        return apiTokenStats;
+    }
+    
+    protected static XmlFile getConfigFile(File parent) {
+        return new XmlFile(new File(parent, "apiTokenStats.xml"));
+    }
+    
+    public static class SingleTokenStats {
+        private final String tokenUuid;
+        private Date lastUseDate;
+        private Integer useCounter;
+        
+        private SingleTokenStats(String tokenUuid) {
+            this.tokenUuid = tokenUuid;
+        }
+        
+        private SingleTokenStats readResolve() {
+            if (this.useCounter != null) {
+                // to avoid negative numbers to be injected
+                this.useCounter = Math.max(0, this.useCounter);
+            }
+            return this;
+        }
+        
+        private void notifyUse() {
+            this.useCounter = useCounter == null ? 1 : useCounter + 1;
+            this.lastUseDate = new Date();
+        }
+        
+        public String getTokenUuid() {
+            return tokenUuid;
+        }
+        
+        // used by Jelly view
+        public int getUseCounter() {
+            return useCounter == null ? 0 : useCounter;
+        }
+        
+        // used by Jelly view
+        public Date getLastUseDate() {
+            return lastUseDate;
+        }
+        
+        // used by Jelly view
+        /**
+         * Return the number of days since the last usage
+         * Relevant only if the lastUseDate is not null
+         */
+        public long getNumDaysUse() {
+            return lastUseDate == null ? 0 : computeDeltaDays(lastUseDate.toInstant(), Instant.now());
+        }
+        
+        private long computeDeltaDays(Instant a, Instant b) {
+            long deltaDays = ChronoUnit.DAYS.between(a, b);
+            deltaDays = Math.max(0, deltaDays);
+            return deltaDays;
+        }
+        
+        private static Comparator<SingleTokenStats> COMP_BY_LAST_USE_THEN_COUNTER =
+                Comparator.comparing(SingleTokenStats::getLastUseDate, Comparator.nullsLast(Comparator.naturalOrder()))
+                        .thenComparing(SingleTokenStats::getUseCounter, Comparator.reverseOrder());
```

Comment:
I don't think you want `Comparator.nullsLast` or `Comparator.reverseOrder()`. My expectation based on `keepLastUpdatedDuplicated` is that you want the comparator to sort the following as A < B < C < D (newest and most used tokens should come last so they overwrite tokens with duplicate uuids in the map), but as it is you get C < B < D < A.
Token A: 0 Uses, Last on `null`
Token B: 2 Uses, Last on 2018-01-05 00:00:00
Token C: 3 Uses, Last on 2018-01-05 00:00:00
Token D: 2 Uses, Last on 2018-02-05 00:00:00

The following comparator might be better:

    Comparator.comparing(SingleTokenStats::getLastUseDate, Comparator.nullsFirst(Comparator.naturalOrder()))
            .thenComparing(SingleTokenStats::getUseCounter);

Reply:
Yeah very good point (even if we are in corner case in reality), I checked only B < D in my unit test. 👍 

I added more unit test to cover this part of the comparison.